grovedb/src/operations/proof.rs

```diff
@@ -92,41 +142,50 @@ impl GroveDb {
         query: PathQuery,
     ) -> Result<([u8; 32], Vec<(Vec<u8>, Vec<u8>)>), Error> {
         let path_slices = query.path.iter().map(|x| x.as_slice()).collect::<Vec<_>>();
+
+        if path_slices.len() < 1 {
+            return Err(Error::InvalidPath("can't verify proof for empty path"));
+        }
+
+        let mut result_set: Vec<(Vec<u8>, Vec<u8>)> = vec![];
         let mut proof_reader = ProofReader::new(proof);
 
-        let merk_proof = proof_reader.read_proof(MERK_PROOF)?;
+        let mut current_limit = query.query.limit;
+        let mut current_offset = query.query.offset;
 
-        let (mut last_root_hash, result_set) = merk::execute_proof(
-            &merk_proof,
-            &query.query.query,
-            query.query.limit,
-            query.query.offset,
-            query.query.query.left_to_right,
-        )
-        .expect("should execute proof");
+        let mut expected_root_hash = GroveDb::execute_subquery_proof(
+            &mut proof_reader,
+            &mut result_set,
+            &mut current_limit,
+            &mut current_offset,
+            query.clone(),
+        )?;
 
-        // Validate the path
+        // validate the path elements are connected
         let mut split_path = path_slices.split_last();
         while let Some((key, path_slice)) = split_path {
             if !path_slice.is_empty() {
-                let merk_proof = proof_reader.read_proof(MERK_PROOF)?;
+                // for every subtree, we should have a corresponding proof for the parent
+                // which should prove that this subtree is a child of the parent tree
+                let parent_merk_proof =
+                    proof_reader.read_proof_of_type(ProofType::MerkProof.into())?;
 
                 let mut parent_query = Query::new();
                 parent_query.insert_key(key.to_vec());
 
-                let proof_result = merk::execute_proof(
-                    &merk_proof,
+                let proof_result = execute_merk_proof(
+                    &parent_merk_proof,
                     &parent_query,
                     None,
                     None,
                     query.query.query.left_to_right,
-                )
-                .expect("should execute proof");
-                let result_set = proof_result.1.result_set;
+                )?;
 
-                if result_set[0].0 != key.to_vec() {
+                let result_set = proof_result.1.result_set;
+                if result_set.len() == 0 || result_set[0].0 != key.to_vec() {
```

Comment:
will it work if we take slice from `result_set[0].0` rather than making a vector out of `key` ?

Reply:
fixed in #103